task7-LedDriver/hello.c

```diff
@@ -0,0 +1,165 @@
+#include <linux/init.h> // Macros used to mark up functions e.g., __init __exit
+#include <linux/module.h> // Core header for loading LKMs into the kernel
+#include <linux/kernel.h> // Contains types, macros, functions for the kernel
+#include <linux/proc_fs.h>
+#include <linux/uaccess.h>
+#include <linux/string.h>
+#include <linux/gpio.h>
+#include <linux/timer.h>
+#include <linux/jiffies.h>
+#include <linux/err.h>
+
+MODULE_LICENSE("GPL"); ///< The license type -- this affects runtime behavior
+MODULE_AUTHOR("Maksym Khomenko"); ///< The author -- visible when you use modinfo
+MODULE_DESCRIPTION("Simple gpio driver with proc_fs support"); ///< The description -- see modinfo
+MODULE_VERSION("0.2"); ///< The version of the module
+
+#define GPIO_21 (21)		// LED connected to this gpio
+#define GPIO_4 (4)	// Button connected to this gpio
+#define TIMEOUT 500    // 2000 ms
+#define BUTTON_SCAN_PERIOD 100 // 100 ms
+#define PROC_BUFFER_SIZE 256
+
+static char procfs_buffer[PROC_BUFFER_SIZE] = "MODULE NAME: my_module\nAUTHOR: Maksym Khomenko\nDESCRIPTION: Simple gpio driver with proc_fs support\nVERSION: 0.2\nLED STATE: ON\n";
+static size_t procfs_buffer_size = PROC_BUFFER_SIZE;
+static struct proc_dir_entry *proc_file;
+static struct proc_dir_entry *proc_folder;
+
+static struct timer_list etx_timer;		// LED blink timer
+static struct timer_list b_timer;		// Button scan timer
+
+static u8 blink = 1;					// LED state (1 - blink, 0 - not blink)
+
+#define PROC_FILE_NAME "my_info"
+#define PROC_DIR_NAME "my_module" /* /proc/my_module/my_state */
+
+static ssize_t my_module_read(struct file *File, char __user* buffer, size_t count, loff_t *offset)
+{
+	ssize_t to_copy, not_copied, delta;
+
+	if (procfs_buffer_size == 0) 
+		{
+			procfs_buffer_size = strlen(procfs_buffer);
+			return 0;
+		}
+
+	to_copy = min(count, procfs_buffer_size);
+	not_copied = copy_to_user(buffer, procfs_buffer, to_copy);

```

Comment:
Data is copied always from the beginning of the buffer.
Please see my comment in the same location in task6-hello-module-procfs/hello.c